linting/src/primitive_topic.rs

```diff
@@ -0,0 +1,222 @@
+// Copyright 2018-2023 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use clippy_utils::{
+    diagnostics::span_lint_and_then,
+    is_lint_allowed,
+    match_def_path,
+    source::snippet_opt,
+};
+use if_chain::if_chain;
+use rustc_errors::Applicability;
+use rustc_hir::{
+    self,
+    def::{
+        DefKind,
+        Res,
+    },
+    def_id::DefId,
+    Arm,
+    AssocItemKind,
+    ExprKind,
+    Impl,
+    ImplItemKind,
+    ImplItemRef,
+    Item,
+    ItemKind,
+    Node,
+    PatKind,
+    QPath,
+};
+use rustc_lint::{
+    LateContext,
+    LateLintPass,
+};
+use rustc_middle::ty::{
+    self,
+    Ty,
+    TyKind,
+};
+use rustc_session::{
+    declare_lint,
+    declare_lint_pass,
+};
+
+declare_lint! {
+    /// **What it does:** Checks for ink! contracts that use
+    /// the [`#[ink(topic)]`](https://use.ink/macros-attributes/topic) annotation with primitive
+    /// number types. Topics are discrete events for which it makes sense to filter. Typical
+    /// examples of fields that should be filtered are `AccountId`, `bool` or `enum` variants.
+    ///
+    /// **Why is this bad?** It typically doesn't make sense to annotate types like `u32` or `i32`
+    /// as a topic, if those fields can take continuous values that could be anywhere between
+    /// `::MIN` and `::MAX`. An example of a case where it doesn't make sense at all to have a
+    /// topic on the storage field is something like `value: Balance` in the examle below.
+    ///
+    /// **Known problems:** Events 2.0 currently are not supported:
```

Comment:
If events are not supported then we should either merge this PR into `4.2` branch or make it compatible with the Events 2.0

Reply:
My bad, it looks like your recent commit: https://github.com/paritytech/ink/pull/1837/commits/f10985bd17393a43f48468be305fd9801a1e1696 added the support for the Events 2.0. If that's the case, can you please update the comment?